### ISSUE TITLE
Warcraft III (& The Frozen Throne) - crash on resolution change

### ISSUE THREAD

**Issue by WinterSnowfall** [AlpyneDreams/d8vk#142]
As the titles says, the games crash on resolution change with d8vk, while working fine with WineD3D. The game window resizes to the desired (newly set) resolution, but then the game crashes.

The problem doesn't seem to be captured in the traces, as the game will completely unload d3d8 and then reload it again to switch resolution. This is why you'll actually see two trace files included below. 

The crash with d8vk occurs after the first trace is done and right before anything can be captured in the trace corresponding to the new resolution, so I believe it might be the same initialization problem we see with other games such as #59 or with 3DMark2001.

### Software information
Warcraft III + The Frozen Throne (Retail, patched up to 1.26a - the last version to use d3d8).

### System information
- GPU: RTX 3060
- Driver: 525.105.17
- Wine version: 8.6 Staging
- DXVK version: ed3a395

### Apitrace file(s)
[war3.trace..tar.xz](https://drive.google.com/file/d/1QO3eztVvmih7OM-vGv0dpM3ckPWou3Bz/view?usp=share_link)

### Log files
- d3d8.log: [war3_d3d8.log](https://github.com/AlpyneDreams/d8vk/files/11322017/war3_d3d8.log)

**Comment by WinterSnowfall** [comment]
Performance might also be something to look into as there's a constant queue sync while in game and the frametimes are somewhat stuttery. Could just be how the game does things though, since it's not very noticable and everything seems smooth actually.

![image](https://user-images.githubusercontent.com/6306593/234514628-8aa4fc16-b12f-4059-9dda-f6cf062a04e1.png)

From some of my tests WineD3D performs better on my Intel Xe, managing around 60 fps @ 1080p, while with d8vk I get about 40 fps @ 1080p. I haven't noticed any difference on Nvidia, but the game runs capped at 60 fps anyway, so it might just be outside of the measurable range on a 3060.

**Comment by WinterSnowfall** [comment]
Thanks to @K0bin ([see this](https://github.com/K0bin/dxvk/commit/8a66b987fa2840c575d9d61ed2f94fcc6895dae1)), I've looked a bit into what is causing that queue sync.

`warn:  Waiting for resource: Lock Image, pool: 0 Usage: 1 Flags: 2064 Flags: 2064 Format: D3D9Format::X8R8G8B8`

If I add the following at the start of D3D9DeviceEx::LockImage():

`if (unlikely(pResource->Desc()->Usage & D3DUSAGE_RENDERTARGET)) return D3DERR_INVALIDCALL;`

That gets rid of the queue sync with apparently no ill effect (other than the game being less choppy as a result).

@K0bin mentioned Sonic Generations also relied on this call to fail. It might have been a thing back in the Win98/XP era perhaps? I doubt we can get a trace to verify it, so :man_shrugging: .

Not sure if we necessarily want/need to address this, but thought I'd mention it anyway.

**Comment by AlpyneDreams** [comment]
Testing this on Windows and it looks like it might be something to do with `D3D9WindowProc`. Somehow it ends up recursing infinitely into itself.

![image](https://user-images.githubusercontent.com/3376691/236611909-e446a5d0-90bd-4107-a4f9-7acb85f0f24f.png)

Probably an upstream issue, seems like a headache. I've tried merging in the latest from upstream, but that didn't fix it. It's possible the behavior on Wine might be different but I imagine it's probably the same issue.

**Comment by WinterSnowfall** [comment]
Well, that's unfortunate, but at least we can rule out any issues with winevulkan now.

**Comment by K0bin** [comment]
Wonder if this is the same thing I stumbled across years ago: https://github.com/doitsujin/dxvk/issues/1695

**Comment by AlpyneDreams** [comment]
Okay, after a lot more testing I've found the issue. Turns out somehow we are leaking refs to the d3d9 device. So the window proc is never unhooked even after the d3d8 device is freed. I'm not sure where those refs are coming from but I'm looking into it now.

**Comment by AlpyneDreams** [comment]
Okay, I've found all sources of extra refs, and I've fixed an upstream bug. Now the game should work.

**Comment by AlpyneDreams** [comment]
> Thanks to @K0bin ([see this](https://github.com/K0bin/dxvk/commit/8a66b987fa2840c575d9d61ed2f94fcc6895dae1)), I've looked a bit into what is causing that queue sync.
> 
> `warn: Waiting for resource: Lock Image, pool: 0 Usage: 1 Flags: 2064 Flags: 2064 Format: D3D9Format::X8R8G8B8`
> 
> If I add the following at the start of D3D9DeviceEx::LockImage():
> 
> `if (unlikely(pResource->Desc()->Usage & D3DUSAGE_RENDERTARGET)) return D3DERR_INVALIDCALL;`
> 
> That gets rid of the queue sync with apparently no ill effect (other than the game being less choppy as a result).
> 
> @K0bin mentioned Sonic Generations also relied on this call to fail. It might have been a thing back in the Win98/XP era perhaps? I doubt we can get a trace to verify it, so 🤷‍♂️ .
> 
> Not sure if we necessarily want/need to address this, but thought I'd mention it anyway.

The game creates the device with `D3DPRESENTFLAG_LOCKABLE_BACKBUFFER` so unfortunately we have to let it do this.

**Comment by K0bin** [comment]
That makes things very simple. Game just stupid.

**Comment by dungeon007** [comment]
As i see there is an option in registry called **lockfb** so i guess if somebody dislike this, could set it to 0 and voila.
Maybe that is what @WinterSnowfall wanted?

**Comment by WinterSnowfall** [comment]
Nice catch. I've set that to 0 and indeed the queue sync is gone now. Makes you wonder why they left it on 1 for release though? Maybe some cards had issues with it disabled?

Anyway, I just finished the base game with it enabled - guess I can disable it while playing The Frozen Throne and see if that causes any issues. If not, I'll add a note to PCGamingWiki about it, assuming I get @dungeon007's blessing :smile:.
